Update SSR doc to reflect streaming (#2851)
withastro/docs: src/content/docs/en/guides/server-side-rendering.mdx

### Server Endpoints

A server endpoint, also known as an **API route**, is a special function exported from a `.js` or `.ts` file within the `src/pages/` folder. 
The function takes an [endpoint context](/en/reference/api-reference/#endpoint-context) and returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). A powerful feature of SSR, API routes are able to securely execute code on the server. To learn more, see our [Endpoints Guide](/en/core-concepts/endpoints/#server-endpoints-api-routes).

### Streaming

> Since browsers support HTTP streaming, a document will be broken up into chunks, sent over the network in order, and rendered on the page in that order.

Reviewer: I would rephrase this from the perspective of what the browser does. Something like:

> Browsers consume HTML incrementally; parsing, rendering into the DOM, and painting. This happens as the full HTML document is consumed. This happens whether or not you intentionally stream your HTML. Network conditions can cause large documents to be downloaded slowly.

> Astro additionally uses streaming to help "unblock" data fetching. Before this feature, Astro would wait until all data finished loading before sending HTML. Now, Astro will render and send components over the network until it finds one that `await`s a Promise. It will wait for the component's Promises to resolve, then render and send the component over the network. 

Reviewer 2: > The wording in this paragraph with "... before this feature ..." and "Now, Astro will...", reads like a paragraph referencing a new release. 

Considering Astro's supported this for a [while](https://github.com/withastro/astro/releases/tag/astro%401.0.0-beta.55) now, is it worth rephrasing to something along the lines of: 

> ..."Astro's supported streaming in its templates since version (x.x.x). (continue paragraph as usual while committing "Before this feature" and "Now, ..") ... Astro additionally uses streaming to help "unblock"  ...

Arguably, we may remove the version details and just keep the paragraph without that info ie., as the docs largely reflect the current state of the project (v2 docs) 

> Astro additionally uses streaming to help "unblock" data fetching. Instead of waiting until all data finishes loading before sending HTML, Astro will render and send components over the network until it finds one that `await`s a Promise. It will wait for the component's Promises to resolve, then render and send the component over the network.

> Astro additionally uses streaming to help "unblock" data fetching. Before this feature, Astro would wait until all data finished loading before sending HTML. Now, Astro will render and send components over the network until it finds one that `await`s a Promise. It will wait for the component's Promises to resolve, then render and send the component over the network. 

Reviewer: I would probably describe how this works "without" streaming rather than saying "before". I don't think a before/after is very interesting to anyone who wasn't around in the "before" and it becomes increasingly irrelevant as time goes on. But the difference in approaches in the approaches themselves is eternal.

> Astro additionally uses streaming to help "unblock" data fetching. Before this feature, Astro would wait until all data finished loading before sending HTML. Now, Astro will render and send components over the network until it finds one that `await`s a Promise. It will wait for the component's Promises to resolve, then render and send the component over the network. 

Reviewer: Astro doesn't send components over the network at all. I'm confused by this sentence slightly.

### Streaming

Browsers natively support HTTP streaming, where a document is broken up into chunks, sent over the network in order, and rendered on the page in that order.

During this process, browsers consume HTML incrementally: parsing, rendering into the DOM, and painting. This happens whether or not you intentionally stream your HTML. Network conditions can cause large documents to be downloaded slowly, and waiting for data fetches can block page rendering.

### Using streaming to improve page performance

The following page `await`s some data in its frontmatter. Astro will wait for all of the `fetch` calls to resolve before sending any HTML to the browser.

```astro title="src/pages/index.astro"
---
const personResponse = await fetch('https://randomuser.me/api/');
const personData = await personResponse.json();
const randomPerson = personData.results[0];
const factResponse = await fetch('https://catfact.ninja/fact');
const factData = await factResponse.json();
---
<html>
  <head>
    <title>A name and a fact</title>
  </head>
  <body>
    <h2>A name</h2>
    <p>{randomPerson.name.first}</p>
    <h2>A fact</h2>
    <p>{factData.fact}</p>
  </body>
</html>
```

> Moving the `await` calls into smaller components allows you to take advantage of Astro's streaming. Using the following components, Astro can render some HTML first, such as the title, then send the paragraphs when the data is ready.

Reviewer 3: ```suggestion
Moving the `await` calls into smaller components allows you to take advantage of Astro's streaming. Using the following components to perform the data fetches, Astro can render some HTML first, such as the title, and then the paragraphs when the data is ready.
```

```astro title="src/components/RandomName.astro"
---
const personResponse = await fetch('https://randomuser.me/api/');
const personData = await personResponse.json();
const randomPerson = personData.results[0];
---
<p>{randomPerson.name.first}</p>
```

```astro title="src/components/RandomFact.astro"
---
const factResponse = await fetch('https://catfact.ninja/fact');
const factData = await factResponse.json();
---
<p>{factData.fact}</p>
```

> ```astro title="src/pages/index.astro"
> ---
> import RandomName from '../components/RandomName.astro'
> import RandomFact from '../components/RandomFact.astro'
> ---
> <html>
>   <head>
>     <title>A name and a fact</title>
>   </head>
>   <body>
>     <h2>A name</h2>
>     <RandomName />
>     <h2>A fact</h2>
>     <RandomFact />
>   </body>
> </html>
> ```
> 
> The page will now send each part of the page whenever it is ready. Astro will first send the `head`, `body`, and `h1` tags to the browser. It will then wait for `RandomName` to finish loading its Promises, and then send the resulting HTML over the network. It will render the next heading, and then it will wait for `RandomFact` and send its resulting HTML over the browser.

Reviewer 3: ```suggestion

The Astro page below using these components can render parts of the page sooner. The `<head`>, `<body>`, and `<h1>` tags are no longer blocked by data fetches. The browser will then wait for `RandomName` to finish loading its Promises, and then send the resulting HTML over the network. It will render the next heading, and then it will wait for `RandomFact` and send its resulting HTML over the browser.

``````astro title="src/pages/index.astro"
---
import RandomName from '../components/RandomName.astro'
import RandomFact from '../components/RandomFact.astro'
---
<html>
  <head>
    <title>A name and a fact</title>
  </head>
  <body>
    <h2>A name</h2>
    <RandomName />
    <h2>A fact</h2>
    <RandomFact />
  </body>
</html>
```
``````
Broke up the code samples by swapping the order of explanatory paragraph and `index.astro` file that uses the above components

### Server Endpoints

A server endpoint, also known as an **API route**, is a special function exported from a `.js` or `.ts` file within the `src/pages/` folder. 
The function takes an [endpoint context](/en/reference/api-reference/#endpoint-context) and returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). A powerful feature of SSR, API routes are able to securely execute code on the server. To learn more, see our [Endpoints Guide](/en/core-concepts/endpoints/#server-endpoints-api-routes).

### Streaming

Browsers natively support HTTP streaming, where a document is broken up into chunks, sent over the network in order, and rendered on the page in that order.

> During this process, browsers consume HTML incrementally: parsing, rendering into the DOM, and painting. This happens whether or not you intentionally stream your HTML. Network conditions can cause large documents to be downloaded slowly.

Reviewer 3: ```suggestion
During this process, browsers consume HTML incrementally: parsing, rendering into the DOM, and painting. This happens whether or not you intentionally stream your HTML. Network conditions can cause large documents to be downloaded slowly, and waiting for data fetches can block page rendering.
```

> Astro additionally uses streaming to help "unblock" data fetching. Before this feature, Astro would wait until all data finished loading before sending HTML. Now, Astro will render and send components over the network until it finds one that `await`s a Promise. It will wait for the component's Promises to resolve, then render and send the component over the network. 

Reviewer 3: ```suggestion
```
deleting this in favour of starting with a briefer mention in the paragraph above

### Server Endpoints

A server endpoint, also known as an **API route**, is a special function exported from a `.js` or `.ts` file within the `src/pages/` folder. 
The function takes an [endpoint context](/en/reference/api-reference/#endpoint-context) and returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). A powerful feature of SSR, API routes are able to securely execute code on the server. To learn more, see our [Endpoints Guide](/en/core-concepts/endpoints/#server-endpoints-api-routes).

### Streaming

Astro enables HTTP streaming for all pages that are not prerendered. This means that your page's content will be sent incrementally to the page as it resolves async functions and requests.

> You can `await` promises in the frontmatter of your page, or you can move those requests down into Astro components the page renders. The page will then send down the HTML it's capable of sending down without waiting for those async components, and then it will start rendering the components in the order they are found in your page's HTML one-by-one.

Reviewer 4: Nit: no hyphens in "one by one"
```suggestion
You can `await` promises in the frontmatter of your page, or you can move those requests down into Astro components the page renders. The page will then send down the HTML it's capable of sending down without waiting for those async components, and then it will start rendering the components in the order they are found in your page's HTML one by one.
```

For example, given a page that fetches some data:

```astro title="src/pages/index.astro"

> ---
> const data = await fetch('https://example.com')
> const otherData = await fetch('https//other-example.com')
> ---
> <html>
>   <head>
>     <title>Astro example</title>
>   </head>
>   <body>
>     <p>{data.example}</p>
>     <p>{otherData.example}</p>
>   </body>
> </html>

Reviewer 4: Generally, we tend to be a bit more concrete with our code examples, what do you think of making this one functional? 
```suggestion
---
const personResponse = await fetch('https://randomuser.me/api/');
const personData = await personResponse.json();
const randomPerson = personData.results[0];
const factResponse = await fetch('https://catfact.ninja/fact');
const factData = await factResponse.json();
---
<html>
  <head>
    <title>A name and a fact</title>
  </head>
  <body>
    <p>{randomPerson.name.first}</p>
    <p>{factData.fact}</p>
  </body>
</html>
```

Author: LOVE IT @Yan-Thomas

```

We can move the fetch call into a component to allow the page to send down the shell of the markup (the html, head, and body) before the fetch request is done loading:

> ```astro title="src/components/Example.astro"
> ---
> const data = await fetch('https://example.com')
> ---
> <p>{data.example}<p>
> ```

Reviewer 4: ````suggestion
```astro title="src/components/RandomName.astro"
---
const personResponse = await fetch('https://randomuser.me/api/');
const personData = await personResponse.json();
const randomPerson = personData.results[0];
---
<p>{randomPerson.name.first}</p>
```
````

Author: 👍🏽

> ```astro title="src/components/OtherExample.astro"
> ---
> const otherData = await fetch('https://other-example.com')
> ---
> <p>{otherData.example}<p>
> ```

Reviewer 4: ````suggestion
```astro title="src/components/RandomFact.astro"
---
const factResponse = await fetch('https://catfact.ninja/fact');
const factData = await factResponse.json();
---
<p>{factData.fact}</p>
```
````

Author: 💯

And update our page to stream in the component instead:

```astro title="src/pages/index.astro" {1, 9}

> ---
> import Example from '../components/Example.astro'
> import OtherExample from '../components/OtherExample.astro'
> ---
> <html>
>   <head>
>     <title>Astro example</title>
>   </head>
>   <body>
>     <Example />
>     <OtherExample />
>   </body>
> </html>

Reviewer 4: ```suggestion
---
import RandomName from '../components/RandomName.astro'
import RandomFact from '../components/RandomFact.astro'
---
<html>
  <head>
    <title>A name and a fact</title>
  </head>
  <body>
    <RandomName />
    <RandomFact />
  </body>
</html>
```

Author: 🥇

### Server Endpoints

A server endpoint, also known as an **API route**, is a special function exported from a `.js` or `.ts` file within the `src/pages/` folder. 
The function takes an [endpoint context](/en/reference/api-reference/#endpoint-context) and returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). A powerful feature of SSR, API routes are able to securely execute code on the server. To learn more, see our [Endpoints Guide](/en/core-concepts/endpoints/#server-endpoints-api-routes).

### Streaming

Astro enables HTTP streaming for all pages that are not prerendered. This means that your page's content will be sent incrementally to the page as it resolves async functions and requests.

> You can `await` promises in the frontmatter of your page, or you can move those requests down into Astro components the page renders. The page will then send down the HTML it's capable of sending down without waiting for those async components, and then it will start rendering the components in the order they are found in your page's HTML one by one.

Reviewer 3: ```suggestion
To control the order in which your page renders content, you can adjust where you `await` promises. 

When you `await` promises in the frontmatter of your page, your initial page rendering waits for this async data. Moving those requests into separate Astro components means that your page around these components can now render without waiting. Then, it will start rendering the async components in the order they are found in your page's HTML, one by one.
```
K, just a rough edit to start and I might have lost some nuance (and might have got something wrong!), but I think something like this hits the use-case harder up front?

> For example, given a page that fetches some data:

Reviewer 3: ```suggestion
For example, the following page fetches some data in its frontmatter, which delays page rendering:
```

```astro title="src/pages/index.astro"
---
const personResponse = await fetch('https://randomuser.me/api/');
const personData = await personResponse.json();
const randomPerson = personData.results[0];
const factResponse = await fetch('https://catfact.ninja/fact');
const factData = await factResponse.json();
---
<html>
  <head>
    <title>A name and a fact</title>
  </head>
  <body>
    <p>{randomPerson.name.first}</p>
    <p>{factData.fact}</p>
  </body>
</html>
```

> We can move the fetch call into a component to allow the page to send down the shell of the markup (the html, head, and body) before the fetch request is done loading:
> 
> ```astro title="src/components/RandomName.astro"
> ---
> const personResponse = await fetch('https://randomuser.me/api/');
> const personData = await personResponse.json();
> const randomPerson = personData.results[0];
> ---
> <p>{randomPerson.name.first}</p>
> ```
> 
> ```astro title="src/components/RandomFact.astro"
> ---
> const factResponse = await fetch('https://catfact.ninja/fact');
> const factData = await factResponse.json();
> ---
> <p>{factData.fact}</p>
> ```
> 
> And update our page to stream in the component instead:
> 
> ```astro title="src/pages/index.astro" {1, 9}
> ---
> import RandomName from '../components/RandomName.astro'
> import RandomFact from '../components/RandomFact.astro'
> ---
> <html>
>   <head>
>     <title>A name and a fact</title>
>   </head>
>   <body>
>     <RandomName />
>     <RandomFact />
>   </body>
> </html>
> ```
> 
> Now the page will stream in the HTML around the async components first, and then resolve each component one by one, streaming the resulting HTML.

Reviewer 3: ``````suggestion
To take advantage of Astro's HTML streaming, move the fetch calls into separate components included in the template. This allows the page to send down the shell of the markup (the `<html>`, `<head>`, and `<body>`) before the fetch request is done loading:

```astro title="src/pages/index.astro" {1, 9}
---
import RandomName from '../components/RandomName.astro'
import RandomFact from '../components/RandomFact.astro'
---
<html>
  <head>
    <title>A name and a fact</title>
  </head>
  <body>
    <RandomName />
    <RandomFact />
  </body>
</html>
```
Now the page will stream in the HTML around the async components first, and then resolve each component (shown below) one by one, streaming the resulting HTML.

```astro title="src/components/RandomName.astro"
---
const personResponse = await fetch('https://randomuser.me/api/');
const personData = await personResponse.json();
const randomPerson = personData.results[0];
---
<p>{randomPerson.name.first}</p>
```

```astro title="src/components/RandomFact.astro"
---
const factResponse = await fetch('https://catfact.ninja/fact');
const factData = await factResponse.json();
---
<p>{factData.fact}</p>
```
``````
I like what you've both done with the example here! (I'm half wondering whether this is now a recipe, but...)

I'm not enen *entirely* convinced we need to show the code for the individual components here, to be honest! (That's why my mind jumped to recipe.) What I did here was reorder to show essentially what we're trying to demonstrate: the comparison of what a page looks like when its fetch requests are in frontmatter (causing delays) and what a page looks like when streaming async components can happen. That *might* be all that's necessary here, to get the point across! Not sure, but take a look at how this page now reads with the subtle reordering and see what you think!

Author: @sarah11918 these edits are amazing!

I'll wait on the platform team's review to make sure we've got this right before committing to it.
